### ISSUE TITLE
[Issue]: AutoencoderKLFlux2/AutoencoderOobleck decode is non-deterministic and numerically wrong on gfx1101 (torch 2.13+rocm7.14) while every primitive is bit-deterministic

### ISSUE THREAD

**Issue by linxuhao** [ROCm/ROCm#6633]
### Problem Description

`diffusers.AutoencoderKLFlux2.decode()` (FLUX.2-klein-4B VAE) returns a **different, mutually
uncorrelated result on every call** on gfx1101 (RX 7800 XT), in a single process, with the
identical input tensor and the identical weights, under `torch.no_grad()`. The same weights on
CPU give a stable, correct result. `AutoencoderOobleck.decode()` (Stable Audio Open 1.0) fails
the same way.

At the same time, **every primitive these modules are built from is bit-deterministic on the
same device** and numerically accurate against CPU, and the GPU RNG is fine. So the failure is
in the composition, not in any single op.

#### Decisive cross-check: the same model on the same GPU is correct via Vulkan

The same FLUX.2-klein-4B model, the same GPU, run through ggml/Vulkan (RADV NAVI32) with
stable-diffusion.cpp produces a **correct 512x512 image in 15.81 s total**, of which the VAE
decode — the exact operation that is broken under ROCm — took **0.68 s**.

This rules out faulty hardware, a bad model, and an unsupported configuration. The defect is
specific to the ROCm software path.

#### Non-determinism: five consecutive `decode()` calls, one process, identical input

| run | std | range |
|---|---|---|
| 0 | 0.24787 | [-1.926, +1.568] |
| 1 | 0.13946 | [-1.095, +0.710] |
| 2 | 0.24474 | [-1.722, +1.301] |
| 3 | 0.22251 | [-1.235, +1.111] |
| 4 | 0.23901 | [-1.025, +1.300] |

bit-identical across runs: **False** (expected: True)

| comparison | SNR |
|---|---|
| run0 vs run1 | -1.2 dB |
| run0 vs run2 | -0.6 dB |
| run0 vs run3 | -0.2 dB |
| run0 vs run4 | -0.3 dB |

A negative SNR means the runs are not merely non-bit-exact — they are uncorrelated with each other.

Reference, same weights on CPU: std = 0.14060, range = [-1.071, +0.631] (stable and correct).

Additional failure modes on the same stack:

| case | result |
|---|---|
| `enable_vae_tiling()` on GPU | output explodes to range [-291980, +60727], SNR -94.9 dB vs CPU |
| `AutoencoderOobleck.decode()` (Stable Audio Open 1.0) | GPU std 0.00309 vs CPU std 0.06870, SNR -0.0 dB (amplitude ~22x too small, uncorrelated) |
| MusicGen EnCodec decoder, **same GPU** | correct: GPU vs CPU SNR **+47.5 dB** |

#### Every primitive is bit-deterministic and accurate

5 runs each, same device, same inputs — max diff 0.000e+00 in all cases:

| primitive | bit-deterministic (5 runs) |
|---|---|
| Conv2d 512->512 3x3 | yes (max diff 0.000e+00) |
| GroupNorm(32) | yes (max diff 0.000e+00) |
| SiLU | yes (max diff 0.000e+00) |
| interpolate x2 nearest | yes (max diff 0.000e+00) |
| SDPA | yes (max diff 0.000e+00) |
| GEMM 2048x2048 | yes (max diff 0.000e+00) |
| Conv1d 128->128 k7 | yes (max diff 0.000e+00) |
| weight_norm Conv1d | yes (max diff 0.000e+00) |

Accuracy of those same primitives vs CPU: fp32 +118..+131 dB, fp16 +59..+69 dB — all fine.

GPU RNG is fine as well: `torch.randn` on cuda gives mean -0.00365, std 0.99909, 199714 unique values.

The pattern — composition non-deterministic, every constituent primitive bit-deterministic — is
consistent with uninitialized memory or a race. I have not localised it further and am not
claiming a specific root cause.

#### Already ruled out

Please do not close this as one of the following; each was tested:

- **Not an `HSA_OVERRIDE_GFX_VERSION` problem.** Reproduces with the override **unset** (native
  gfx1101), and also with `11.0.0` and with `11.0.1`. Non-deterministic in all three.
- **Not GPU-reset state.** Reproduces on a clean boot with **zero** prior GPU resets (boot 13:54,
  image broken at 13:58; the first reset of that boot was at 14:19).
- **Not the kernel.** Upgrading 7.0.0-28 -> 7.0.0-29 changed nothing.
- **Not a stale kernel cache.** Clearing the Triton and MIOpen caches changed nothing.
- **Not the weights.** All 248 VAE parameter tensors are finite and non-degenerate, and the same
  weights decode correctly on CPU.
- **Not `PYTORCH_CUDA_ALLOC_CONF=expandable_segments`.** Disabling it does not help.
- **Not the hardware and not the model.** See the Vulkan cross-check above: same GPU, same model,
  correct output.

#### Observation (not a diagnosis)

`torch.cuda.get_arch_list()` returns only `['gfx942']` on this install, even though the
`amd-torch-device-gfx1101` and `rocm-sdk-device-gfx1101` wheels are installed (modular
device-wheel packaging). Reporting this in case it is relevant; I am not claiming it is the cause.

### Operating System

Ubuntu 24.04.4 LTS (Noble Numbat), kernel 7.0.0-29-generic

### CPU

AMD Ryzen 7 7800X3D 8-Core Processor

### GPU

AMD Radeon RX 7800 XT (gfx1101, Navi 32, 16 GB) — affected. A second GPU in the same box, a
Radeon RX 7900 XTX (gfx1100) running vLLM, is unaffected.

### ROCm Version

ROCm 7.14 (hip 7.14.60850), torch 2.13.0+rocm7.14.0

Container image: `rocm/pytorch@sha256:a3867e22ca44c458b1feda92026c6dc289e7e447093cf55914d1fd56ef6d7372`

### ROCm Component

Not determined. The failing path is PyTorch (torch 2.13.0+rocm7.14.0) `decode()` on gfx1101;
the individual primitives it is built from all pass, so I cannot attribute it to a single
component.

### Steps to Reproduce

Run the script below inside
`rocm/pytorch@sha256:a3867e22ca44c458b1feda92026c6dc289e7e447093cf55914d1fd56ef6d7372` on a
gfx1101 GPU. Part 1 shows every primitive is bit-deterministic; part 2 shows five consecutive
`decode()` calls on the identical input are not.

```python
"""
Minimal repro: AutoencoderKLFlux2.decode() is NON-DETERMINISTIC on ROCm.

Five consecutive decode() calls with the identical input tensor return
uncorrelated results, while every primitive the module is built from is
bit-deterministic on the same device.

Env: torch 2.13.0+rocm7.14.0, hip 7.14.60850, AMD RX 7800 XT (gfx1101),
     image rocm/pytorch@sha256:a3867e22ca44c458b1feda92026c6dc289e7e447093cf55914d1fd56ef6d7372
Reproduces with HSA_OVERRIDE_GFX_VERSION unset (native gfx1101), and also
with 11.0.0 / 11.0.1.
"""
import torch, numpy as np, torch.nn.functional as F
from diffusers import Flux2KleinPipeline

print(f"torch={torch.__version__} hip={torch.version.hip} "
      f"device={torch.cuda.get_device_properties(0).gcnArchName}")

# ---- 1. every primitive used by the module is bit-deterministic ----
def det(name, fn, *ts, n=5):
    outs = []
    for _ in range(n):
        with torch.no_grad():
            outs.append(fn(*ts).float().cpu().numpy().copy())
    print(f"  {name:28s} bit-deterministic={all(np.array_equal(outs[0], o) for o in outs[1:])}")

x = torch.randn(1, 512, 32, 32).cuda(); w = torch.randn(512, 512, 3, 3).cuda() * 0.02
det("Conv2d 512->512 3x3", lambda a, b: F.conv2d(a, b, padding=1), x, w)
det("GroupNorm(32)", lambda a: F.group_norm(a, 32), x)
det("SiLU", F.silu, x)
det("interpolate x2 nearest", lambda a: F.interpolate(a, scale_factor=2, mode="nearest"), x)

# ---- 2. the module composed from them is NOT ----
vae = Flux2KleinPipeline.from_pretrained(
    "black-forest-labs/FLUX.2-klein-4B", torch_dtype=torch.float32).vae.eval()
torch.manual_seed(0)
z = torch.randn(1, 32, 64, 64)
with torch.no_grad():
    ref = vae.to("cpu").decode(z).sample.float().numpy()
vae = vae.to("cuda"); zc = z.cuda()

outs = []
for i in range(5):
    with torch.no_grad():
        y = vae.decode(zc).sample.float().cpu().numpy().copy()
    outs.append(y)
    print(f"  run {i}: std={y.std():.5f} range=[{y.min():+.3f},{y.max():+.3f}]")

print(f"\nbit-identical across runs: {all(np.array_equal(outs[0], o) for o in outs[1:])}  <-- expected True")
for i in range(1, 5):
    a, b = outs[0].astype(np.float64), outs[i].astype(np.float64)
    print(f"  run0 vs run{i}: SNR={10*np.log10(np.sum(a**2)/max(np.sum((a-b)**2),1e-30)):+6.1f} dB")
a, b = ref.astype(np.float64), outs[0].astype(np.float64)
print(f"GPU vs CPU: SNR={10*np.log10(np.sum(a**2)/max(np.sum((a-b)**2),1e-30)):+6.1f} dB   "
      f"(CPU std={ref.std():.5f}, GPU std={outs[0].std():.5f})")
```

### Additional Information

- Measured 2026-08-20.
- `AutoencoderOobleck` (Stable Audio Open 1.0) reproduces the same failure with a 1-D decoder
  stack, so the defect is not specific to the 2-D convolution path.
- MusicGen's EnCodec decoder on the same GPU in the same environment is correct (+47.5 dB vs
  CPU), so it is not "any decoder stack on gfx1101".
- Stable Audio hung this GPU twice during testing, requiring a reset. The non-determinism above
  is reproducible without any prior reset (see the ruled-out list).
